src/python/espressomd/cuda_init.pyx

```diff
@@ -52,35 +51,72 @@ cdef class CudaInitHandle:
                 Set the device id of the graphics card to use.
 
             """
-            cuda_set_device(_dev)
+            cuda_set_device(dev)
 
-    IF CUDA == 1:
-        @property
-        def device_list(self):
+        def list_devices(self):
             """
             List devices.
 
             Returns
             -------
-            :obj:`list` :
+            :obj:`dict` :
                 List of available CUDA devices.
 
             """
             cdef char gpu_name_buffer[4 + 64]
+            n_gpus = 0
+            try:
+                n_gpus = cuda_get_n_gpus()
+            except RuntimeError:
+                pass
             devices = dict()
-            for i in range(cuda_get_n_gpus()):
-                cuda_get_gpu_name(i, gpu_name_buffer)
-                devices[i] = gpu_name_buffer
+            for i in range(n_gpus):
+                try:
+                    cuda_get_gpu_name(i, gpu_name_buffer)
+                except RuntimeError:
+                    continue
+                devices[i] = utils.to_str(gpu_name_buffer)
             return devices
 
-        @device_list.setter
-        def device_list(self, dict _dev_dict):
-            raise Exception("cuda device list is read only")
+        def list_devices_properties(self):
+            """
+            List devices with their properties on each host machine.
 
+            Returns
+            -------
+            :obj:`dict` :
+                List of available CUDA devices with their properties.
+
+            """
+            cdef vector[EspressoGpuDevice] devices
+            cdef EspressoGpuDevice dev
+            try:
+                devices = cuda_gather_gpus()
+            except RuntimeError:
+                pass
+            resources = dict()
+            for i in range(devices.size()):
+                dev = devices[i]
+                hostname = utils.to_str(dev.proc_name)
+                if hostname not in resources:
+                    resources[hostname] = {}
+                resources[hostname][dev.id] = {
```

Comment:
can't we get the device as a string directly?

Reply:
The string returned by `cudaGetDeviceProperties()` cannot be used as a key: we only collect the first 64 characters of the string returned by the CUDA function. I'm not sure why we truncate it, it was written like this 6 years ago and the commit message doesn't say anything, and there is not comment to explain it. Maybe back different CUDA versions had different char array lengths?

Reply:
i meant to get a string representation of ```cudaDeviceProp``` right in the core instead of building it on the python level

Reply:
Sorry, I don't get what you mean. On this line we're building a dict. Most of the device properties the user is interested in are integers.

Reply:
oh, I was missing the dict, thought we only want to print the information, sorry